### PR TITLE
00852 d new account balance export

### LINKED ISSUE #852: Export a new *_Balances.proto file to (eventually) supersede current *_Balances.csv
**Summary**
The account balances CSV has an awkward format for token balances; it gives all an account's token balances in a single column via the Base64 encoding of a serialized `TokenBalances` protobuf message,
```
message TokenBalance {
    TokenID tokenId = 1;
    uint64 balance = 2; 
}

message TokenBalances {
    repeated TokenBalance tokenBalances = 1;
}
```

**Planned resolution**
Create a new (second) pure protobuf file that captures the entire contents of the current CSV file via a serialized `AllAccountBalances` message,
Continue writing  current *_Balances.csv file.

```
message TokenBalance {
   TokenID tokenId = 1;
   uint64 balance = 2;
}

message SingleAccountBalances {
   AccountID accountID = 1;
   uint64 hbarBalance = 2;
   repeated TokenBalance tokenBalances = 3;
}

message AllAccountBalances {
   Timestamp consensusTimestamp = 1;
   repeated SingleAccountBalances allAccounts = 2;
}
```

Tasks:
- [x] Add the protobuf file exporting;
- [x] Add account balances validator to validate the exported account balances:
    - [x] compare the all account info of .csv and the .proto file from one node;
    - [x] compare all account info of .proto files from all nodes;

### PR DESCRIPTION
**Pull request by ljianghedera**
**Related issue(s)**:
Closes #852

Slack links for JRSL https://hedera-hashgraph.slack.com/archives/C018CBQBTGB/p1612927392276600 

**Summary of the change**:
1. Add the AllAccountBalances protobuf message exporting;
2. Need to add account balances validators as well;

**External impacts**:
None.

**Applicable documentation**
- [ ] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 1]
```diff
@@ -124,6 +132,37 @@ public void toCsvFile(ServicesState signedState, Instant when) {
 		if (exportSucceeded) {
 			tryToSign(csvLoc);
 		}
+		log.info("It took total {} Millisecond to export and sign the csv account files", (System.nanoTime() - startTime) / 1_000_000);
+	}
+
+	// TODO: add the log info to tell how long it take to export and sign the proto account balances file
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:138]: This TODO can be removed , as we are logging the info needed to export in line 165

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 2]
```diff
@@ -96,6 +102,7 @@ public SignedStateBalancesExporter(
 	public boolean isTimeToExport(Instant now) {
 		if (periodEnd == NEVER) {
 			periodEnd = now.plusSeconds(dynamicProperties.balancesExportPeriodSecs());
+			return false;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:105]: This return is redundant as line 112 returns the same.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 3]
```diff
@@ -124,6 +132,37 @@ public void toCsvFile(ServicesState signedState, Instant when) {
...
 			tryToSign(csvLoc);
 		}
+		log.info("It took total {} Millisecond to export and sign the csv account files", (System.nanoTime() - startTime) / 1_000_000);
+	}
+
+	// TODO: add the log info to tell how long it take to export and sign the proto account balances file
+	//       as some validation info
+	//       Do the same to csv file.
+	@Override
+	public void toProtoFile(ServicesState signedState, Instant when) {
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:142]: nit : `when` can be renamed to `timeStamp` or similar

File: hedera-node/src/test/java/com/hedera/services/state/exports/SignedStateBalancesExporterTest.java [hunk 4]
```diff
@@ -304,18 +306,77 @@ public boolean shouldExportTokenBalances() {
...
+
+
+	private String expectedBalancesName(Boolean isProto ) {
+		return isProto ? now.toString().replace(":", "_") + "_Balances.proto"
+				: now.toString().replace(":", "_") + "_Balances.csv";
 	}
 
 	private String expectedBalancesName() {
-		return now.toString().replace(":", "_") + "_Balances.csv";
+		return  expectedBalancesName(false) ; // now.toString().replace(":", "_") + "_Balances.csv";
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/test/java/com/hedera/services/state/exports/SignedStateBalancesExporterTest.java:376]: nit : commented code can be removed

File: pom.xml [hunk 5]
```diff
@@ -138,13 +137,13 @@
...
     <maven-versions.version>2.8.1</maven-versions.version>
     <commons-io.version>2.1</commons-io.version>
-
     <!-- Dependency properties -->
     <commons-codec.version>1.11</commons-codec.version>
     <eddsa.version>0.3.0</eddsa.version>
     <grpc.version>1.33.0</grpc.version>
     <jackson-databind.version>2.9.10.7</jackson-databind.version>
-    <javax.annotation-api.version>1.3.2</javax.annotation-api.version>
+    <javax.annotation-api.version>
```
Comment (Neeharika-Sompalli) [inline pom.xml:145]: nit: I believe due to code alignment these lines moved to next line. These changes can be reverted for readability.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 6]
```diff
@@ -124,6 +132,37 @@ public void toCsvFile(ServicesState signedState, Instant when) {
...
+					when,total,	expectedFloat));
+		}
+
+		var protoLoc = lastUsedExportDir + when.toString().replace(":", "_") + "_Balances.proto";
+		boolean exportSucceeded = exportBalancesProtoFile(allAccountBalancesBuilder, protoLoc);
+
+		if (exportSucceeded) {
+			tryToSign(protoLoc);
+		}
+		log.info("It took total {} Millisecond to export and sign the proto account files", (System.nanoTime() - startTime) / 1_000_000);
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:165]: `1_000_000` can be used as `Units.NANOSECONDS_TO_MILLISECONDS`

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 7]
```diff
@@ -124,6 +132,37 @@ public void toCsvFile(ServicesState signedState, Instant when) {
 		if (exportSucceeded) {
 			tryToSign(csvLoc);
 		}
+		log.info("It took total {} Millisecond to export and sign the csv account files", (System.nanoTime() - startTime) / 1_000_000);
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:135]: Same here please use `Units.NANOSECONDS_TO_MILLISECONDS`

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 2 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:105]: I was thinking to let it return `true` here to force it to do an export asap it restarts. I will change it that way.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 1 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:138]: Yes, done.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 2 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:105]: Renamed.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 3 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:142]: Renamed

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 7 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:135]: Done.

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java [hunk 8]
```diff
@@ -165,6 +204,82 @@ private boolean exportBalancesFile(BalancesSummary summary, String csvLoc, Insta
...
+		try {
+			FileInputStream fin = new FileInputStream(protoLoc);
+			AllAccountBalances allAccountBalances = AllAccountBalances.parseFrom(fin);
+			return Optional.ofNullable(allAccountBalances);
+		} catch (InvalidProtocolBufferException e) {
+			log.error("protobuf message file is corrupted: {}", protoLoc);
+		} catch (IOException e) {
+			log.error("Can't read protobuf message file {}", protoLoc);
+		}
+		return Optional.empty();
```
Comment (anighanta) [inline hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:280]: can we add test cases such that, given AllAccountBalances.parseFrom(fin) will throw InvalidProtocolBufferException and IOException and validate we return empty AllAccountBalances

**Review by anighanta — COMMENTED** [review]
codecov test handling reports can be resolved by adding tests. LGTM otherwise

File: hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java (hunk 8 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/state/exports/SignedStateBalancesExporter.java:280]: Yeah, I will.

**Review by Neeharika-Sompalli — COMMENTED** [review]
Please merge `master` into the branch before merging. Apart from changes to `pom.xml` rest LGTM.

**Review by Neeharika-Sompalli — DISMISSED** [review]

**Review by Neeharika-Sompalli — APPROVED** [review]

File: pom.xml [hunk 9]
```diff
@@ -316,6 +316,8 @@ ${git.signoff}
           <artifactId>maven-surefire-plugin</artifactId>
           <version>${maven-surefire.version}</version>
           <configuration>
+            <forkCount>3</forkCount>
```
Comment (Neeharika-Sompalli) [inline pom.xml:319]: Why do we need these new configuration?

Comment (ljianghedera) [reply pom.xml:319]: That is to fix a weird test failure: 

[INFO] Hedera Services Node ............................... FAILURE [12:55 min]
[INFO] Hedera Services Test Clients ....................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  13:08 min
[INFO] Finished at: 2021-02-10T04:19:08Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.2:test (default-test) on project hedera-node: There are test failures.
[ERROR] 
[ERROR] Please refer to /repo/hedera-node/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] There was an error in the forked process
[ERROR] Java heap space
[ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: There was an error in the forked process
[ERROR] Java heap space
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.fork(ForkStarter.java:656)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:282)
[ERROR] 	at org.apache.maven.plugin.surefire.booterclient.ForkStarter.run(ForkStarter.java:245)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeProvider(AbstractSurefireMojo.java:1183)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.executeAfterPreconditionsChecked(AbstractSurefireMojo.java:1011)
[ERROR] 	at org.apache.maven.plugin.surefire.AbstractSurefireMojo.execute(AbstractSurefireMojo.java:857)
[ERROR] 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
[ERROR] 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
[ERROR] 	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
[ERROR] 	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
[ERROR] 	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
[ERROR] 	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
[ERROR] 	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:956)
[ERROR] 	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
[ERROR] 	at org.apache.maven.cli.MavenCli.main(MavenCli.java:192)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
[ERROR] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.

**Pull request merged**